## Practica 3 

#### Maldicion de dimensionalidad


In [5]:
import numpy as np
from sklearn.preprocessing import PolynomialFeatures

In [31]:
values_2_3 = np.arange(6).reshape(2,3)
values_3_6 = np.arange(18).reshape(3,6)
values_3_9 = np.arange(27).reshape(3,9)


**Polinomio grado 1**

In [18]:
pol1 = PolynomialFeatures(1)
pol1

PolynomialFeatures(degree=1, include_bias=True, interaction_only=False,
                   order='C')

In [33]:
# solo genera el intercepto y cada variable linealmente
pol1.fit_transform(values_2_3)

array([[1., 0., 1., 2.],
       [1., 3., 4., 5.]])

**Polinomio grado 2**

In [20]:
pol2 = PolynomialFeatures(2)
pol2

PolynomialFeatures(degree=2, include_bias=True, interaction_only=False,
                   order='C')

In [34]:
# genera cada combinacion de valores
# el input es [a,b,c] , [d,e,f]
# intecepto, a, b, c, aa, ab, ac, bb, bc, cc
# intecepto, d, e, f, dd, de, df, ee, ef, ff
pol2.fit_transform(values_2_3)

array([[ 1.,  0.,  1.,  2.,  0.,  0.,  0.,  1.,  2.,  4.],
       [ 1.,  3.,  4.,  5.,  9., 12., 15., 16., 20., 25.]])

**Polinomio grado 3**

In [29]:
pol3 = PolynomialFeatures(3)
pol3

PolynomialFeatures(degree=3, include_bias=True, interaction_only=False,
                   order='C')

In [30]:
# intercepto, a, b, c, aa, ab, ac, bb, bc, cc, aaa, aab, aac, abb, abc, acc, bbb, bbc, bcc, ccc
pol3.fit_transform(values_2_3)

array([[  1.,   0.,   1.,   2.,   0.,   0.,   0.,   1.,   2.,   4.,   0.,
          0.,   0.,   0.,   0.,   0.,   1.,   2.,   4.,   8.],
       [  1.,   3.,   4.,   5.,   9.,  12.,  15.,  16.,  20.,  25.,  27.,
         36.,  45.,  48.,  60.,  75.,  64.,  80., 100., 125.]])

In [36]:
pol3.fit_transform(values_2_3).shape

(2, 20)

#### Utilizando 6 variables para cada grado de generacion de polinomios

In [32]:
#
pol1.fit_transform(values_3_6)

array([[ 1.,  0.,  1.,  2.,  3.,  4.,  5.],
       [ 1.,  6.,  7.,  8.,  9., 10., 11.],
       [ 1., 12., 13., 14., 15., 16., 17.]])

In [37]:
pol2.fit_transform(values_3_6)

array([[  1.,   0.,   1.,   2.,   3.,   4.,   5.,   0.,   0.,   0.,   0.,
          0.,   0.,   1.,   2.,   3.,   4.,   5.,   4.,   6.,   8.,  10.,
          9.,  12.,  15.,  16.,  20.,  25.],
       [  1.,   6.,   7.,   8.,   9.,  10.,  11.,  36.,  42.,  48.,  54.,
         60.,  66.,  49.,  56.,  63.,  70.,  77.,  64.,  72.,  80.,  88.,
         81.,  90.,  99., 100., 110., 121.],
       [  1.,  12.,  13.,  14.,  15.,  16.,  17., 144., 156., 168., 180.,
        192., 204., 169., 182., 195., 208., 221., 196., 210., 224., 238.,
        225., 240., 255., 256., 272., 289.]])

In [38]:
pol3.fit_transform(values_3_6)

array([[1.000e+00, 0.000e+00, 1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00,
        5.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00, 5.000e+00,
        4.000e+00, 6.000e+00, 8.000e+00, 1.000e+01, 9.000e+00, 1.200e+01,
        1.500e+01, 1.600e+01, 2.000e+01, 2.500e+01, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
        0.000e+00, 1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00, 5.000e+00,
        4.000e+00, 6.000e+00, 8.000e+00, 1.000e+01, 9.000e+00, 1.200e+01,
        1.500e+01, 1.600e+01, 2.000e+01, 2.500e+01, 8.000e+00, 1.200e+01,
        1.600e+01, 2.000e+01, 1.800e+01, 2.400e+01, 3.000e+01, 3.200e+01,
        4.000e+01, 5.000e+01, 2.700e+01, 3.600e+01, 4.500e+01, 4.800e+01,
        6.000e+01, 7.500e+01, 6.400e+0

In [39]:
print(pol1.fit_transform(values_3_6).shape)
print(pol2.fit_transform(values_3_6).shape)
print(pol3.fit_transform(values_3_6).shape)

(3, 7)
(3, 28)
(3, 84)


#### Utilizando 9 variables

In [40]:
print(pol1.fit_transform(values_3_9).shape)
print(pol2.fit_transform(values_3_9).shape)
print(pol3.fit_transform(values_3_9).shape)

(3, 10)
(3, 55)
(3, 220)


## Conclusiones

Debido a que se realiza una combinacion de cada variable segun el grado de polinomio, las combinaciones crecen significativamente segun estos valores.  Como se logra observar para un grado de polinomio 1 solo se tienen los valores lineales de cada caso, pero cuando se realiza un polinomio de grado 2 se calculan las combinaciones de cada variable mas los cuadrados.

Por lo que tener varias variables genera un polinomio grande como en el ultimo caso, donde con 9 variables un polinomio de grado 3 genera 220 factores.  Este modelo se volveria demasiado grande para determinados casos; por ejemplo si se quisiera utilizar para tratamiento de imagenes. 

Si la imagen fuera pequeña, de 40 x 40, serian 1600 variables; parte de la utilidad de este proceso es saber donde utilizarlo de forma apropiada en donde la cantidad de variables es poca y el grado del polinomio no es tran grande ya que esto podria provocar overfiting.







